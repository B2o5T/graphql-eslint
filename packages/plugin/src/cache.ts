// Based on the `eslint-plugin-import`'s cache
// https://github.com/import-js/eslint-plugin-import/blob/main/utils/ModuleCache.js
import debugFactory from 'debug';

const log = debugFactory('graphql-eslint:ModuleCache');

export class ModuleCache<K, T> {
  map = new Map<K, { lastSeen: [number, number]; result: T }>();

  set(cacheKey: K, result: T): void {
    this.map.set(cacheKey, { lastSeen: process.hrtime(), result });
    log('setting entry for', cacheKey);
  }

  get(cacheKey: K, settings = { lifetime: 10 /* seconds */ }): T | void {
    const value = this.map.get(cacheKey);
    if (!value) {
      log('cache miss for', cacheKey);
      return;
    }
    const { lastSeen, result } = value;
    // check freshness
    if (
      process.env.NODE /* don't check for ESLint CLI */ ||
      process.hrtime(lastSeen)[0] < settings.lifetime
    ) {
      return result;
    }
  }
}
