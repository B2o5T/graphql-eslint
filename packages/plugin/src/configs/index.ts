import { ConfigName } from '../types.js';
// @ts-expect-error -- complains about no default export
import operationsAllConfig from './operations-all.js';
// @ts-expect-error -- complains about no default export
import operationsRecommendedConfig from './operations-recommended.js';
// @ts-expect-error -- complains about no default export
import schemaAllConfig from './schema-all.js';
// @ts-expect-error -- complains about no default export
import schemaRecommendedConfig from './schema-recommended.js';
// @ts-expect-error -- complains about no default export
import relayConfig from './schema-relay.js';

export const configs = {
  'schema-recommended': schemaRecommendedConfig,
  'schema-all': schemaAllConfig,
  'schema-relay': relayConfig,
  'operations-recommended': operationsRecommendedConfig,
  'operations-all': operationsAllConfig,
  'flat/schema-recommended': schemaRecommendedConfig.rules,
  'flat/schema-all': {
    ...schemaRecommendedConfig.rules,
    ...schemaAllConfig.rules,
  },
  'flat/schema-relay': relayConfig.rules,
  'flat/operations-recommended': operationsRecommendedConfig.rules,
  'flat/operations-all': {
    ...operationsRecommendedConfig.rules,
    ...operationsAllConfig.rules,
  },
} satisfies Record<ConfigName | `flat/${ConfigName}`, unknown>;
