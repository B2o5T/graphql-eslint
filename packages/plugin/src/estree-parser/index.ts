export * from "./converter";
export * from "./estree-ast";
export * from "./utils";
