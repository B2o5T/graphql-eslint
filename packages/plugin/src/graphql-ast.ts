import {
  GraphQLSchema,
  GraphQLFieldMap,
  GraphQLInputFieldMap,
  GraphQLField,
  GraphQLInputField,
  GraphQLInputType,
  GraphQLOutputType,
  GraphQLNamedType,
  GraphQLInterfaceType,
  GraphQLArgument,
  isObjectType,
  isInterfaceType,
  isUnionType,
  isInputObjectType,
  isListType,
  isNonNullType,
} from 'graphql';

export function getReachableTypes(schema: GraphQLSchema): Set<string>
export function getReachableTypes(schema?: GraphQLSchema): Set<string> | null {
  if (schema) {
    return collectReachableTypes(schema);
  }
  
  return null;
}

export function collectReachableTypes(schema: GraphQLSchema): Set<string> {
  const reachableTypes = new Set<string>();

  collectFrom(schema.getQueryType());
  collectFrom(schema.getMutationType());
  collectFrom(schema.getSubscriptionType());

  return reachableTypes;

  function collectFrom(type?: GraphQLNamedType): void {
    if (type && shouldCollect(type.name)) {
      if (isObjectType(type) || isInterfaceType(type)) {
        collectFromFieldMap(type.getFields());
        collectFromInterfaces(type.getInterfaces());
      } else if (isUnionType(type)) {
        type.getTypes().forEach(collectFrom);
      } else if (isInputObjectType(type)) {
        collectFromInputFieldMap(type.getFields());
      }
    }
  }

  function collectFromFieldMap(fieldMap: GraphQLFieldMap<any, any>): void {
    for (const fieldName in fieldMap) {
      collectFromField(fieldMap[fieldName]);
    }
  }

  function collectFromField(field: GraphQLField<any, any>): void {
    collectFromOutputType(field.type);
    field.args.forEach(collectFromArgument);
  }

  function collectFromArgument(arg: GraphQLArgument): void {
    collectFromInputType(arg.type);
  }

  function collectFromInputFieldMap(fieldMap: GraphQLInputFieldMap): void {
    for (const fieldName in fieldMap) {
      collectFromInputField(fieldMap[fieldName]);
    }
  }

  function collectFromInputField(field: GraphQLInputField): void {
    collectFromInputType(field.type);
  }

  function collectFromInterfaces(interfaces: GraphQLInterfaceType[]): void {
    if (interfaces) {
      interfaces.forEach(interfaceType => {
        collectFromFieldMap(interfaceType.getFields());
        collectFromInterfaces(interfaceType.getInterfaces());
      });
    }
  }

  function collectFromOutputType(output: GraphQLOutputType): void {
    collectFrom(schema.getType(resolveName(output)));
  }

  function collectFromInputType(input: GraphQLInputType): void {
    collectFrom(schema.getType(resolveName(input)));
  }

  function resolveName(type: GraphQLOutputType | GraphQLInputType) {
    if (isListType(type) || isNonNullType(type)) {
      return resolveName(type.ofType);
    }

    return type.name;
  }

  function shouldCollect(name: string): boolean {
    if (!reachableTypes.has(name)) {
      reachableTypes.add(name);
      return true;
    }

    return false;
  }
}
