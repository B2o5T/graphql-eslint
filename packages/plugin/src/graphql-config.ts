import { dirname } from 'path';
import debugFactory from 'debug';
import {
  GraphQLConfig,
  GraphQLExtensionDeclaration,
  loadConfigSync,
  SchemaPointer,
} from 'graphql-config';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import { ParserOptions } from './types';

const debug = debugFactory('graphql-eslint:graphql-config');
let graphQLConfig: GraphQLConfig;

export function loadGraphQLConfig(options: ParserOptions = {}): GraphQLConfig {
  // We don't want cache config on test environment
  // Otherwise schema and documents will be same for all tests
  if (process.env.NODE_ENV !== 'test' && graphQLConfig) {
    return graphQLConfig;
  }

  const onDiskConfig = options.skipGraphQLConfig
    ? null
    : loadConfigSync({
        // load config relative to the file being linted
        rootDir: options.filePath ? dirname(options.filePath) : undefined,
        throwOnEmpty: false,
        throwOnMissing: false,
        extensions: [addCodeFileLoaderExtension],
      });

  debug('options.skipGraphQLConfig: %o', options.skipGraphQLConfig);
  if (onDiskConfig) {
    debug('Graphql-config path %o', onDiskConfig.filepath);
  }

  const configOptions = options.projects
    ? { projects: options.projects }
    : {
        schema: (options.schema || '') as SchemaPointer, // if `schema` is `undefined` will throw error `Project 'default' not found`
        documents: options.documents || options.operations,
        extensions: options.extensions,
        include: options.include,
        exclude: options.exclude,
      };

  graphQLConfig =
    onDiskConfig ||
    new GraphQLConfig(
      {
        config: configOptions,
        filepath: 'virtual-config',
      },
      [addCodeFileLoaderExtension]
    );

  return graphQLConfig;
}

const addCodeFileLoaderExtension: GraphQLExtensionDeclaration = api => {
  api.loaders.schema.register(new CodeFileLoader());
  api.loaders.documents.register(new CodeFileLoader());
  return { name: 'graphql-eslint-loaders' };
};
