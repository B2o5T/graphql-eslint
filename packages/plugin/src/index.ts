import { configs } from './configs/index.js';
import { parseForESLint, parser } from './parser.js';
import { processor } from './processor.js';
import { rules } from './rules/index.js';

export * from './types.js';
export type { IGraphQLConfig } from 'graphql-config';
export type { GraphQLTagPluckOptions } from '@graphql-tools/graphql-tag-pluck';

export { requireGraphQLSchemaFromContext, requireSiblingsOperations } from './utils.js';

export const processors = { graphql: processor };

export { rules, configs, parser, parseForESLint };

// eslint-disable-next-line import/no-default-export -- It's common practice for ESLint plugins that supports Flat config to use the default export
export default {
  parser,
  processor,
  rules,
  configs,
};
