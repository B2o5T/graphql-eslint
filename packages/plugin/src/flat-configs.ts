import { parseForESLint } from './parser';
import { rules } from './rules';
import { configs } from './configs';

const baseOptions = {
  parser: { parseForESLint },
  plugins: {
    '@graphql-eslint': { rules },
  },
};

export const flatConfigs = {
  'operations-all': {
    ...baseOptions,
    rules: {
      ...configs['operations-recommended'].rules,
      ...configs['operations-all'].rules,
    },
  },
  'operations-recommended': {
    ...baseOptions,
    rules: configs['operations-recommended'].rules,
  },
  relay: {
    ...baseOptions,
    rules: configs.relay.rules,
  },
  'schema-all': {
    ...baseOptions,
    rules: {
      ...configs['schema-recommended'].rules,
      ...configs['schema-all'].rules,
    },
  },
  'schema-recommended': {
    ...baseOptions,
    rules: configs['schema-recommended'].rules,
  },
};
