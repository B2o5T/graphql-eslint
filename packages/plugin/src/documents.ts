import { resolve } from 'path';
import {
  FragmentDefinitionNode,
  Kind,
  OperationDefinitionNode,
  SelectionSetNode,
  visit,
  OperationTypeNode,
} from 'graphql';
import { Source } from '@graphql-tools/utils';
import { GraphQLProjectConfig } from 'graphql-config';
import debugFactory from 'debug';
import fg from 'fast-glob';
import { logger } from './utils.js';
import { Pointer } from './types.js';
import { ModuleCache } from './cache.js';

export type FragmentSource = { filePath: string; document: FragmentDefinitionNode };
export type OperationSource = { filePath: string; document: OperationDefinitionNode };

const debug = debugFactory('graphql-eslint:operations');

export type SiblingOperations = {
  available: boolean;
  getFragment(fragmentName: string): FragmentSource[];
  getFragments(): FragmentSource[];
  getFragmentByType(typeName: string): FragmentSource[];
  getFragmentsInUse(
    baseOperation: OperationDefinitionNode | FragmentDefinitionNode | SelectionSetNode,
    recursive?: boolean,
  ): FragmentDefinitionNode[];
  getOperation(operationName: string): OperationSource[];
  getOperations(): OperationSource[];
  getOperationByType(operationType: OperationTypeNode): OperationSource[];
};

const handleVirtualPath = (documents: Source[]): Source[] => {
  const filepathMap: Record<string, number> = Object.create(null);

  return documents.map(source => {
    const location = source.location!;
    if (['.gql', '.graphql'].some(extension => location.endsWith(extension))) {
      return source;
    }
    filepathMap[location] ??= -1;
    const index = (filepathMap[location] += 1);
    return {
      ...source,
      location: resolve(location, `${index}_document.graphql`),
    };
  });
};

const operationsCache = new ModuleCache<Source[]>();
const siblingOperationsCache = new Map<Source[], SiblingOperations>();

const getSiblings = (project: GraphQLProjectConfig): Source[] => {
  const documentsKey = project.documents;
  if (!documentsKey) {
    return [];
  }

  let siblings = operationsCache.get(documentsKey);

  if (!siblings) {
    debug('Loading operations from %o', project.documents);
    const documents = project.loadDocumentsSync(project.documents, {
      skipGraphQLImport: true,
      pluckConfig: project.extensions.pluckConfig,
    });
    if (debug.enabled) {
      debug('Loaded %d operations', documents.length);
      const operationsPaths = fg.sync(project.documents as Pointer, { absolute: true });
      debug('Operations pointers %O', operationsPaths);
    }
    siblings = handleVirtualPath(documents);
    operationsCache.set(documentsKey, siblings);
  }

  return siblings;
};

export function getDocuments(project: GraphQLProjectConfig): SiblingOperations {
  const siblings = getSiblings(project);

  if (siblings.length === 0) {
    let printed = false;

    const noopWarn = () => {
      if (!printed) {
        logger.warn(
          'getSiblingOperations was called without any operations. Make sure to set "parserOptions.operations" to make this feature available!',
        );
        printed = true;
      }
      return [];
    };

    return {
      available: false,
      getFragment: noopWarn,
      getFragments: noopWarn,
      getFragmentByType: noopWarn,
      getFragmentsInUse: noopWarn,
      getOperation: noopWarn,
      getOperations: noopWarn,
      getOperationByType: noopWarn,
    };
  }

  // Since the siblings array is cached, we can use it as cache key.
  // We should get the same array reference each time we get
  // to this point for the same graphql project
  const value = siblingOperationsCache.get(siblings);

  if (value) {
    return value;
  }

  let fragmentsCache: FragmentSource[] | null = null;

  const getFragments = (): FragmentSource[] => {
    if (fragmentsCache === null) {
      const result: FragmentSource[] = [];

      for (const source of siblings) {
        for (const definition of source.document?.definitions || []) {
          if (definition.kind === Kind.FRAGMENT_DEFINITION) {
            result.push({
              filePath: source.location!,
              document: definition,
            });
          }
        }
      }
      fragmentsCache = result;
    }
    return fragmentsCache;
  };

  let cachedOperations: OperationSource[] | null = null;

  const getOperations = (): OperationSource[] => {
    if (cachedOperations === null) {
      const result: OperationSource[] = [];

      for (const source of siblings) {
        for (const definition of source.document?.definitions || []) {
          if (definition.kind === Kind.OPERATION_DEFINITION) {
            result.push({
              filePath: source.location!,
              document: definition,
            });
          }
        }
      }
      cachedOperations = result;
    }
    return cachedOperations;
  };

  const getFragment = (name: string) => getFragments().filter(f => f.document.name?.value === name);

  const collectFragments = (
    selectable: SelectionSetNode | OperationDefinitionNode | FragmentDefinitionNode,
    recursive: boolean,
    collected = new Map<string, FragmentDefinitionNode>(),
  ) => {
    visit(selectable, {
      FragmentSpread(spread) {
        const fragmentName = spread.name.value;
        const [fragment] = getFragment(fragmentName);

        if (!fragment) {
          logger.warn(
            `Unable to locate fragment named "${fragmentName}", please make sure it's loaded using "parserOptions.operations"`,
          );
          return;
        }
        if (!collected.has(fragmentName)) {
          collected.set(fragmentName, fragment.document);
          if (recursive) {
            collectFragments(fragment.document, recursive, collected);
          }
        }
      },
    });
    return collected;
  };

  const siblingOperations: SiblingOperations = {
    available: true,
    getFragment,
    getFragments,
    getFragmentByType: typeName =>
      getFragments().filter(f => f.document.typeCondition?.name?.value === typeName),
    getFragmentsInUse: (selectable, recursive = true) =>
      Array.from(collectFragments(selectable, recursive).values()),
    getOperation: name => getOperations().filter(o => o.document.name?.value === name),
    getOperations,
    getOperationByType: type => getOperations().filter(o => o.document.operation === type),
  };

  siblingOperationsCache.set(siblings, siblingOperations);
  return siblingOperations;
}
