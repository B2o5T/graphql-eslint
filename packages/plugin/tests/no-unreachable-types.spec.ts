import { GraphQLRuleTester, ParserOptions } from '../src';
import rule from '../src/rules/no-unreachable-types';

const useSchema = (schema: string): { code: string; parserOptions: ParserOptions } => {
  return {
    parserOptions: { schema },
    code: schema,
  };
};

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests('no-unreachable-types', rule, {
  valid: [
    useSchema(/* GraphQL */ `
      scalar A
      scalar B

      # UnionTypeDefinition
      union Response = A | B

      type Query {
        foo: Response
      }
    `),
    useSchema(/* GraphQL */ `
      type Query {
        me: User
      }

      # ObjectTypeDefinition
      type User {
        id: ID
        name: String
      }
    `),
    useSchema(/* GraphQL */ `
      type Query {
        me: User
      }

      # InterfaceTypeDefinition
      interface Address {
        city: String
      }

      type User implements Address {
        city: String
      }
    `),
    useSchema(/* GraphQL */ `
      # ScalarTypeDefinition
      scalar DateTime

      type Query {
        now: DateTime
      }
    `),
    useSchema(/* GraphQL */ `
      # EnumTypeDefinition
      enum Role {
        ADMIN
        USER
      }

      type Query {
        role: Role
      }
    `),
    useSchema(/* GraphQL */ `
      input UserInput {
        id: ID
      }

      type Query {
        # InputValueDefinition
        user(input: UserInput!): Boolean
      }
    `),
    useSchema(/* GraphQL */ `
      # DirectiveDefinition
      directive @auth(role: [Role!]!) on FIELD_DEFINITION

      enum Role {
        ADMIN
        USER
      }

      type Query {
        # Directive
        user: ID @auth(role: [ADMIN])
      }
    `),
    useSchema(/* GraphQL */ `
      type RootQuery
      type RootMutation
      type RootSubscription

      schema {
        query: RootQuery
        mutation: RootMutation
        subscription: RootSubscription
      }
    `),
    useSchema(/* GraphQL */ `
      interface User {
        id: ID!
      }

      interface Manager implements User {
        id: ID!
      }

      type TopManager implements Manager {
        id: ID!
        name: String
      }

      type Query {
        me: User
      }
    `),
    {
      name: 'directive on schema',
      ...useSchema(/* GraphQL */ `
        type Query

        schema @good {
          query: Query
        }

        directive @good on SCHEMA
      `),
    },
  ],
  invalid: [
    {
      ...useSchema(/* GraphQL */ `
        type Query {
          node(id: ID!): AnotherNode!
        }

        interface Node {
          id: ID!
        }

        interface AnotherNode {
          createdAt: String
        }

        interface User implements Node {
          id: ID!
          name: String
        }

        type SuperUser implements User & Node {
          id: ID!
          name: String
          address: String
        }
      `),
      errors: [
        { message: 'Interface type `Node` is unreachable.' },
        { message: 'Interface type `User` is unreachable.' },
        { message: 'Object type `SuperUser` is unreachable.' },
      ],
    },
    {
      ...useSchema(/* GraphQL */ `
        # ScalarTypeDefinition
        scalar DateTime

        # EnumTypeDefinition
        enum Role {
          ADMIN
          USER
        }

        # DirectiveDefinition
        directive @auth(role: [String!]!) on FIELD_DEFINITION

        # UnionTypeDefinition
        union Union = String | Boolean

        # InputObjectTypeDefinition
        input UsersFilter {
          limit: Int
        }

        # InterfaceTypeDefinition
        interface Address {
          city: String
        }

        # ObjectTypeDefinition
        type User implements Address {
          city: String
        }
      `),
      errors: [
        { message: 'Scalar type `DateTime` is unreachable.' },
        { message: 'Enum type `Role` is unreachable.' },
        { message: 'Directive `auth` is unreachable.' },
        { message: 'Union type `Union` is unreachable.' },
        { message: 'Input object type `UsersFilter` is unreachable.' },
        { message: 'Interface type `Address` is unreachable.' },
        { message: 'Object type `User` is unreachable.' },
      ],
    },
    {
      ...useSchema(/* GraphQL */ `
        interface User {
          id: String
        }

        type SuperUser implements User {
          id: String
          superDetail: SuperDetail
        }

        type SuperDetail {
          detail: String
        }

        type Query {
          user: User!
        }

        scalar DateTime
      `),
      errors: [{ message: 'Scalar type `DateTime` is unreachable.' }],
    },
    {
      ...useSchema(/* GraphQL */ `
        interface User {
          id: String
        }

        interface AnotherUser {
          createdAt: String
        }

        type SuperUser implements User {
          id: String
        }

        # ObjectTypeExtension
        extend type SuperUser {
          detail: String
        }

        type Query {
          user: AnotherUser!
        }
      `),
      errors: [
        { message: 'Interface type `User` is unreachable.' },
        { message: 'Object type `SuperUser` is unreachable.' },
        { message: 'Object type `SuperUser` is unreachable.' },
      ],
    },
    {
      ...useSchema(/* GraphQL */ `
        type Query {
          node(id: ID!): Node!
        }

        interface Node {
          id: ID!
        }

        interface User implements Node {
          id: ID!
          name: String
        }

        type SuperUser implements User & Node {
          id: ID!
          name: String
          address: String
        }

        scalar DateTime
      `),
      errors: [{ message: 'Scalar type `DateTime` is unreachable.' }],
    },
  ],
});
