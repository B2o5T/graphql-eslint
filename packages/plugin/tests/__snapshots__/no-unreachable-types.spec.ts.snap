// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Error 1/3

   1 |
   2 |         type Query {
   3 |           node(id: ID!): AnotherNode!
   4 |         }
   5 |
>  6 |         interface Node {
     |                   ^^^^ Type "Node" is unreachable
   7 |           id: ID!
   8 |         }
   9 |
  10 |         interface AnotherNode {
  11 |           createdAt: String
  12 |         }
  13 |
  14 |         interface User implements Node {
  15 |           id: ID!
  16 |           name: String
  17 |         }
  18 |
  19 |         type SuperUser implements User & Node {
  20 |           id: ID!
  21 |           name: String
  22 |           address: String
  23 |         }
  24 |       

Error 2/3

   1 |
   2 |         type Query {
   3 |           node(id: ID!): AnotherNode!
   4 |         }
   5 |
   6 |         interface Node {
   7 |           id: ID!
   8 |         }
   9 |
  10 |         interface AnotherNode {
  11 |           createdAt: String
  12 |         }
  13 |
> 14 |         interface User implements Node {
     |                   ^^^^ Type "User" is unreachable
  15 |           id: ID!
  16 |           name: String
  17 |         }
  18 |
  19 |         type SuperUser implements User & Node {
  20 |           id: ID!
  21 |           name: String
  22 |           address: String
  23 |         }
  24 |       

Error 3/3

   1 |
   2 |         type Query {
   3 |           node(id: ID!): AnotherNode!
   4 |         }
   5 |
   6 |         interface Node {
   7 |           id: ID!
   8 |         }
   9 |
  10 |         interface AnotherNode {
  11 |           createdAt: String
  12 |         }
  13 |
  14 |         interface User implements Node {
  15 |           id: ID!
  16 |           name: String
  17 |         }
  18 |
> 19 |         type SuperUser implements User & Node {
     |              ^^^^^^^^^ Type "SuperUser" is unreachable
  20 |           id: ID!
  21 |           name: String
  22 |           address: String
  23 |         }
  24 |       
`;

exports[` 2`] = `
Error 1/7

   1 |
   2 |         # ScalarTypeDefinition
>  3 |         scalar DateTime
     |                ^^^^^^^^ Type "DateTime" is unreachable
   4 |
   5 |         # EnumTypeDefinition
   6 |         enum Role {
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
  12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
  13 |
  14 |         # UnionTypeDefinition
  15 |         union Union = String | Boolean
  16 |
  17 |         # InputObjectTypeDefinition
  18 |         input UsersFilter {
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
  23 |         interface Address {
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
  28 |         type User implements Address {
  29 |           city: String
  30 |         }
  31 |       

Error 2/7

   1 |
   2 |         # ScalarTypeDefinition
   3 |         scalar DateTime
   4 |
   5 |         # EnumTypeDefinition
>  6 |         enum Role {
     |              ^^^^ Type "Role" is unreachable
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
  12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
  13 |
  14 |         # UnionTypeDefinition
  15 |         union Union = String | Boolean
  16 |
  17 |         # InputObjectTypeDefinition
  18 |         input UsersFilter {
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
  23 |         interface Address {
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
  28 |         type User implements Address {
  29 |           city: String
  30 |         }
  31 |       

Error 3/7

   1 |
   2 |         # ScalarTypeDefinition
   3 |         scalar DateTime
   4 |
   5 |         # EnumTypeDefinition
   6 |         enum Role {
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
> 12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
     |                   ^^^^^ Type "auth" is unreachable
  13 |
  14 |         # UnionTypeDefinition
  15 |         union Union = String | Boolean
  16 |
  17 |         # InputObjectTypeDefinition
  18 |         input UsersFilter {
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
  23 |         interface Address {
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
  28 |         type User implements Address {
  29 |           city: String
  30 |         }
  31 |       

Error 4/7

   1 |
   2 |         # ScalarTypeDefinition
   3 |         scalar DateTime
   4 |
   5 |         # EnumTypeDefinition
   6 |         enum Role {
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
  12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
  13 |
  14 |         # UnionTypeDefinition
> 15 |         union Union = String | Boolean
     |               ^^^^^ Type "Union" is unreachable
  16 |
  17 |         # InputObjectTypeDefinition
  18 |         input UsersFilter {
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
  23 |         interface Address {
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
  28 |         type User implements Address {
  29 |           city: String
  30 |         }
  31 |       

Error 5/7

   1 |
   2 |         # ScalarTypeDefinition
   3 |         scalar DateTime
   4 |
   5 |         # EnumTypeDefinition
   6 |         enum Role {
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
  12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
  13 |
  14 |         # UnionTypeDefinition
  15 |         union Union = String | Boolean
  16 |
  17 |         # InputObjectTypeDefinition
> 18 |         input UsersFilter {
     |               ^^^^^^^^^^^ Type "UsersFilter" is unreachable
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
  23 |         interface Address {
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
  28 |         type User implements Address {
  29 |           city: String
  30 |         }
  31 |       

Error 6/7

   1 |
   2 |         # ScalarTypeDefinition
   3 |         scalar DateTime
   4 |
   5 |         # EnumTypeDefinition
   6 |         enum Role {
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
  12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
  13 |
  14 |         # UnionTypeDefinition
  15 |         union Union = String | Boolean
  16 |
  17 |         # InputObjectTypeDefinition
  18 |         input UsersFilter {
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
> 23 |         interface Address {
     |                   ^^^^^^^ Type "Address" is unreachable
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
  28 |         type User implements Address {
  29 |           city: String
  30 |         }
  31 |       

Error 7/7

   1 |
   2 |         # ScalarTypeDefinition
   3 |         scalar DateTime
   4 |
   5 |         # EnumTypeDefinition
   6 |         enum Role {
   7 |           ADMIN
   8 |           USER
   9 |         }
  10 |
  11 |         # DirectiveDefinition
  12 |         directive @auth(role: [String!]!) on FIELD_DEFINITION
  13 |
  14 |         # UnionTypeDefinition
  15 |         union Union = String | Boolean
  16 |
  17 |         # InputObjectTypeDefinition
  18 |         input UsersFilter {
  19 |           limit: Int
  20 |         }
  21 |
  22 |         # InterfaceTypeDefinition
  23 |         interface Address {
  24 |           city: String
  25 |         }
  26 |
  27 |         # ObjectTypeDefinition
> 28 |         type User implements Address {
     |              ^^^^ Type "User" is unreachable
  29 |           city: String
  30 |         }
  31 |       
`;

exports[` 3`] = `
Error 1/1

   1 |
   2 |         interface User {
   3 |           id: String
   4 |         }
   5 |
   6 |         type SuperUser implements User {
   7 |           id: String
   8 |           superDetail: SuperDetail
   9 |         }
  10 |
  11 |         type SuperDetail {
  12 |           detail: String
  13 |         }
  14 |
  15 |         type Query {
  16 |           user: User!
  17 |         }
  18 |
> 19 |         scalar DateTime
     |                ^^^^^^^^ Type "DateTime" is unreachable
  20 |       
`;

exports[` 4`] = `
Error 1/3

   1 |
>  2 |         interface User {
     |                   ^^^^ Type "User" is unreachable
   3 |           id: String
   4 |         }
   5 |
   6 |         interface AnotherUser {
   7 |           createdAt: String
   8 |         }
   9 |
  10 |         type SuperUser implements User {
  11 |           id: String
  12 |         }
  13 |
  14 |         # ObjectTypeExtension
  15 |         extend type SuperUser {
  16 |           detail: String
  17 |         }
  18 |
  19 |         type Query {
  20 |           user: AnotherUser!
  21 |         }
  22 |       

Error 2/3

   1 |
   2 |         interface User {
   3 |           id: String
   4 |         }
   5 |
   6 |         interface AnotherUser {
   7 |           createdAt: String
   8 |         }
   9 |
> 10 |         type SuperUser implements User {
     |              ^^^^^^^^^ Type "SuperUser" is unreachable
  11 |           id: String
  12 |         }
  13 |
  14 |         # ObjectTypeExtension
  15 |         extend type SuperUser {
  16 |           detail: String
  17 |         }
  18 |
  19 |         type Query {
  20 |           user: AnotherUser!
  21 |         }
  22 |       

Error 3/3

   1 |
   2 |         interface User {
   3 |           id: String
   4 |         }
   5 |
   6 |         interface AnotherUser {
   7 |           createdAt: String
   8 |         }
   9 |
  10 |         type SuperUser implements User {
  11 |           id: String
  12 |         }
  13 |
  14 |         # ObjectTypeExtension
> 15 |         extend type SuperUser {
     |                     ^^^^^^^^^ Type "SuperUser" is unreachable
  16 |           detail: String
  17 |         }
  18 |
  19 |         type Query {
  20 |           user: AnotherUser!
  21 |         }
  22 |       
`;

exports[` 5`] = `
Error 1/1

   1 |
   2 |         type Query {
   3 |           node(id: ID!): Node!
   4 |         }
   5 |
   6 |         interface Node {
   7 |           id: ID!
   8 |         }
   9 |
  10 |         interface User implements Node {
  11 |           id: ID!
  12 |           name: String
  13 |         }
  14 |
  15 |         type SuperUser implements User & Node {
  16 |           id: ID!
  17 |           name: String
  18 |           address: String
  19 |         }
  20 |
> 21 |         scalar DateTime
     |                ^^^^^^^^ Type "DateTime" is unreachable
  22 |       
`;
